gcc/rust/ast/rust-ast-builder.cc

```diff
@@ -17,53 +17,73 @@
 // <http://www.gnu.org/licenses/>.
 
 #include "rust-ast-builder.h"
+#include "rust-ast-full-decls.h"
 #include "rust-ast-full.h"
+#include "rust-expr.h"
+#include "rust-token.h"
 
 namespace Rust {
 namespace AST {
 
 std::unique_ptr<Expr>
-AstBuilder::call (std::unique_ptr<Expr> &&path,
-		  std::vector<std::unique_ptr<Expr>> &&args)
+Builder::literal_string (std::string &&content) const
+{
+  return std::unique_ptr<Expr> (
+    new AST::LiteralExpr (std::move (content), Literal::LitType::STRING,
+			  PrimitiveCoreType::CORETYPE_STR, {}, loc));
+}
+
+std::unique_ptr<Expr>
+Builder::call (std::unique_ptr<Expr> &&path,
+	       std::vector<std::unique_ptr<Expr>> &&args) const
 {
   return std::unique_ptr<Expr> (
     new CallExpr (std::move (path), std::move (args), {}, loc));
 }
 
 std::unique_ptr<Expr>
-AstBuilder::identifier (std::string name)
+Builder::array (std::vector<std::unique_ptr<Expr>> &&members) const
+{
+  auto elts = std::unique_ptr<ArrayElems> (
```

Comment:
here we can't do that, because `make_unique` is only templated on `T` - so creating a `ArrayElemsValues` but returning a `std::unique_ptr<ArrayElems>` (base class) does not work. maybe we can work around this but I remember it being annoying in other places. I'll try again here

Reply:
I made this pattern work in other parts of the code. You need to template make_unique with the underlying type and change `auto` to the receiving type.

Reply:
Done in f3644a833682e50e87fc8f39e45010dabbebb47e